mhkit/acoustics/analysis.py

```diff
@@ -0,0 +1,806 @@
+
+    if not isinstance(c, (int, float)):
+        raise TypeError("'c' must be a numeric type (int or float).")
+    if not isinstance(c_seabed, (int, float)):
+        raise TypeError("'c_seabed' must be a numeric type (int or float).")
+
+    if np.any(water_depth <= 0):
+        raise ValueError("All elements of 'water_depth' must be positive numbers.")
+    if c <= 0:
+        raise ValueError("'c' must be a positive number.")
+    if c_seabed <= 0:
+        raise ValueError("'c_seabed' must be a positive number.")
+    if c_seabed <= c:
+        raise ValueError("'c_seabed' must be greater than 'c'.")
+
+    fmin = c / (4 * water_depth * np.sqrt(1 - (c / c_seabed) ** 2))
+
+    return fmin
+
+
+def sound_pressure_spectral_density(
+    pressure: xr.DataArray, fs: Union[int, float], bin_length: Union[int, float] = 1
+) -> xr.DataArray:
+    """
+    Calculates the mean square sound pressure spectral density from audio
+    samples split into FFTs with a specified bin length in seconds, using Hanning
+    windowing with 50% overlap. The amplitude of the PSD is adjusted
+    according to Parseval's theorem.
+
+    Parameters
+    ----------
+    pressure: xarray.DataArray (time)
+        Sound pressure in [Pa] or voltage [V]
+    fs: int or float
+        Data collection sampling rate [Hz]
+    bin_length: int or float
+        Length of time in seconds to create FFTs. Default: 1.
+
+    Returns
+    -------
+    spsd: xarray.DataArray (time, freq)
+        Spectral density [Pa^2/Hz] indexed by time and frequency
+    """
+
+    # Type checks
+    if not isinstance(pressure, xr.DataArray):
+        raise TypeError("'pressure' must be an xarray.DataArray.")
+    if not isinstance(fs, (int, float)):
+        raise TypeError("'fs' must be a numeric type (int or float).")
+    if not isinstance(bin_length, (int, float)):
+        raise TypeError("'bin_length' must be a numeric type (int or float).")
+
+    # Ensure that 'pressure' has a 'time' coordinate
+    if "time" not in pressure.dims:
+        raise ValueError("'pressure' must be indexed by 'time' dimension.")
+
+    # window length of each time series
+    nbin = bin_length * fs
+
+    # Use dolfyn PSD
+    binner = VelBinner(n_bin=nbin, fs=fs, n_fft=nbin)
+    # Always 50% overlap if numbers reshape perfectly
+    # Mean square sound pressure
+    psd = binner.power_spectral_density(pressure, freq_units="Hz")
+    samples = binner.reshape(pressure.values) - binner.mean(pressure.values)[:, None]
+    # Power in time domain
+    t_power = np.sum(samples**2, axis=1) / nbin
+    # Power in frequency domain
+    f_power = psd.sum("freq") * (fs / nbin)
+    # Adjust the amplitude of PSD according to Parseval's theorem
+    psd_adj = psd * t_power[:, None] / f_power
+
+    out = xr.DataArray(
+        psd_adj,
+        coords={"time": psd_adj["time"], "freq": psd_adj["freq"]},
+        attrs={
+            "units": pressure.units + "^2/Hz",
```

Comment:
Are pressure units enforced to be defined somewhere else? 

Maybe we use `getattr` and assign a default null value otherwise?
`units = getattr(pressure, 'units', '') + '^2/Hz'`

Reply:
I manually hardcode them in the io.py file, but if a user wants to use their own imported data they'll run into more issues than just this, I think.